api/agent/protocol/json.go

```diff
@@ -0,0 +1,133 @@
+package protocol
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// This is sent into the function
+// All HTTP request headers should be set in env
+type JSONIO struct {
+	Headers    http.Header `json:"headers,omitempty"`
+	Body       string      `json:"body"`
+	StatusCode int         `json:"status_code,omitempty"`
+}
+
+// JSONProtocol converts stdin/stdout streams from HTTP into JSON format.
+type JSONProtocol struct {
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *JSONProtocol) IsStreamable() bool {
+	return true
+}
+
+type RequestEncoder struct {
+	*json.Encoder
+}
+
+func (h *JSONProtocol) DumpJSON(w io.Writer, req *http.Request) error {
+	stdin := json.NewEncoder(h.in)
+	_, err := io.WriteString(h.in, `{`)
+	if err != nil {
+		// this shouldn't happen
+		return err
+	}
+	if req.Body != nil {
+		_, err := io.WriteString(h.in, `"body":"`)
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		bb := new(bytes.Buffer)
+		_, err = bb.ReadFrom(req.Body)
+		if err != nil {
+			return err
+		}
+
+		err = stdin.Encode(bb.String())
+		if err != nil {
+			return err
+		}
+		_, err = io.WriteString(h.in, `",`)
+		if err != nil {
+			// this shouldn't happen
+			return err
+		}
+		defer bb.Reset()
+		defer req.Body.Close()
```

Comment:
if we are going to do this, it will be in the caller, as there are many dispatch methods and all of them assume consuming the `req.Body`. conveniently, for an http server, we don't need to call this, however I think in the async case we do need to. still, in caller imo